Day4/src/pipeline/extract_employe_data.py

```diff
@@ -0,0 +1,355 @@
+        user ="postgres", 
+        password ="password", 
+        port =5432
+        )
+
+def extract_employee_data(cur,con):
+    delete="DELETE FROM raw_employee"
+    cur.execute(delete)
+    con.commit()
+    with open('../sql/extract_employee.sql') as file:
+        sql = " ".join(file.readlines())
+        cur.execute(sql)
+        con.commit()
+
+def extract_timesheet_data(cur,con):
+    delete="DELETE FROM raw_timesheet"
+    cur.execute(delete)
+    con.commit()
+    with open('../sql/extract_timesheet.sql') as file:
+        sql = " ".join(file.readlines())
+        cur.execute(sql)
+        con.commit()
+
+def clean_hours_worked(cur,con):
+    update_hours_worked="UPDATE raw_timesheet SET hours_worked = '0' WHERE paycode ='ABSENT';"
+    cur.execute(update_hours_worked)
+    con.commit()
+
+def create_temporary_punch_details(cur,con):
+    delete="DROP TABLE  IF EXISTS  temporary_punch_details ;"
+    cur.execute(delete)
+    con.commit()
+
+    create_query='''CREATE TABLE temporary_punch_details AS
+        SELECT DISTINCT(employee_id),punch_apply_date as shift_date, MIN(CAST (punch_in_time as TIMESTAMP)) as shift_start_time, 
+        MAX(CAST(punch_out_time as TIMESTAMP)) as shift_end_time
+        FROM raw_timesheet
+        GROUP BY employee_id, punch_apply_date
+        ORDER by 1 asc, 2 asc;'''
+    cur.execute(create_query)
+    con.commit()
+
+def create_temporary_shift_time(cur,con):
+    delete="DROP TABLE  IF EXISTS  temporary_shift_table ;"
+    cur.execute(delete)
+    con.commit()
+    create_query="""
+        CREATE TABLE temporary_shift_table as
+        SELECT employee_id,shift_date,shift_start_time,shift_end_time,
+        CASE 
+        WHEN shift_start_time :: TIMESTAMP :: time BETWEEN '05:00' AND '11:00':: TIME 
+        THEN 'Morning'
+        WHEN shift_start_time :: TIMESTAMP :: time >= '12:00':: TIME
+        THEN 'Evening'
+        ELSE NULL
+        END as shift_type 
+        FROM temporary_punch_details;"""
+    cur.execute(create_query)
+    con.commit()
+
+def create_temporary_tables_for_hours_worked(cur,con):
+    delete="DROP TABLE  IF EXISTS  temporary_hours_table ;"
+    delete2="DROP TABLE  IF EXISTS  temporary_ahours_table;"
+    cur.execute(delete)
+    cur.execute(delete2)
+    con.commit()
+
+    #temporary table for hours worked as paycode =WRK
+    create_query="""
+    CREATE TABLE temporary_hours_table as
+    SELECT DISTINCT(employee_id),punch_apply_date, SUM(CAST(hours_worked as NUMERIC )) as hours_worked
+    FROM raw_timesheet 
+    WHERE paycode ='WRK'
+    GROUP BY employee_id, punch_apply_date,paycode
+    ORDER BY 1 asc, 2 asc;
+    """
+    #temporary table for hours worked as paycode =ABSENT
+    create_query2="""
+    CREATE TABLE temporary_ahours_table as
+    SELECT DISTINCT(employee_id),punch_apply_date, SUM(CAST(hours_worked as NUMERIC )) as hours_worked
+    FROM raw_timesheet 
+    WHERE paycode ='ABSENT'
+    GROUP BY employee_id, punch_apply_date,paycode
```

Comment:
Is it optimal to create temporary tables for each paycode? `GROUP BY employee_id, punch_apply_date,paycode` is being repeated for each table. A better approach would be to use a single temp table and calculate all the values based on condition. You can look into CASE statements, IIF etc.

Reply:
IT isn't but I had got some errors when using CASE for each paycode